{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "49b72984-5b7a-4c47-bfb9-52cf366d2e92",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘zoo’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    as.Date, as.Date.numeric\n",
      "\n",
      "\n",
      "Loading required package: ensembldb\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: GenomicFeatures\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘AnnotationDbi’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "Loading required package: AnnotationFilter\n",
      "\n",
      "\n",
      "Attaching package: 'ensembldb'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'GGally':\n",
      "  method from   \n",
      "  +.gg   ggplot2\n",
      "\n",
      "Need specific help about ggbio? try mailing \n",
      " the maintainer or visit https://lawremi.github.io/ggbio/\n",
      "\n",
      "\n",
      "Attaching package: 'ggbio'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:ggplot2':\n",
      "\n",
      "    geom_bar, geom_rect, geom_segment, ggsave, stat_bin, stat_identity,\n",
      "    xlim\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'stringr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:ggbio':\n",
      "\n",
      "    fixed\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(dplyr)\n",
    "library(zoo)\n",
    "library(ggplot2)\n",
    "library(seqminer)\n",
    "library(EnsDb.Hsapiens.v86)\n",
    "library(ggbio)\n",
    "library(patchwork)\n",
    "library(stringr)\n",
    "\n",
    "plot_region_pileups = function(chrom, # chromosome\n",
    "                                start, # start coordinate\n",
    "                                end, # end coordinate\n",
    "                                bam_bed_file, # tabix indexed BED file with chrom,start,end,cell\n",
    "                                grouping, # dataframe with cell,group,total_reads\n",
    "                                ensdb=EnsDb.Hsapiens.v86, # EnsDb to allow display of genes\n",
    "                                smoothing_window=100, # window to use for averaging\n",
    "                                x_steps=3000) { # limits number of actual to plot data for\n",
    "\n",
    "  # Validate the grouping dataframe\n",
    "  if (! all(c('cell', 'group', 'total_reads') %in% colnames(grouping))) {\n",
    "    stop('Grouping DF must have cell, grouping, and total_reads columns.')\n",
    "  }\n",
    "\n",
    "  # Get reads for the requested region of BED file (using tabix)\n",
    "  range_string = paste0(chrom, ':', start, '-', end)\n",
    "  reads = seqminer::tabix.read.table(tabixRange=range_string, tabixFile=bam_bed_file)\n",
    "  colnames(reads) = c('chrom', 'start', 'stop', 'cell', 'number')\n",
    "\n",
    "  # Get count of cells per group for normalization\n",
    "  grouping = grouping %>%\n",
    "    group_by(group) %>%\n",
    "    mutate(cells_in_group = n()) %>%\n",
    "    ungroup()\n",
    "\n",
    "  # Restrict to cell barcodes specified by user\n",
    "  reads = dplyr::filter(reads, cell %in% grouping$cell)\n",
    "  reads = dplyr::inner_join(reads, grouping, by='cell')\n",
    "\n",
    "  # Expand reads so have one line per base\n",
    "  expanded_ranges = dplyr::bind_rows(lapply(1:nrow(reads), function(i) {\n",
    "    interval = reads[i, 'start']:reads[i, 'stop']\n",
    "    return(data.frame(position=interval,\n",
    "                      value=1,\n",
    "                      cell=reads[i, 'cell'],\n",
    "                      group=reads[i, 'group'],\n",
    "                      scaling_factor=reads[i, 'total_reads'] ,\n",
    "                      cells_in_group=reads[i, 'cells_in_group']))\n",
    "  }))\n",
    "\n",
    "  # Apply scaling factor and account for different cell counts\n",
    "  # Also apply windowed smoothing\n",
    "  expanded_ranges = expanded_ranges %>%\n",
    "    group_by(position, group) %>%\n",
    "    summarize(total=sum(value / scaling_factor) / mean(cells_in_group)) %>%\n",
    "    group_by(group) %>%\n",
    "    arrange(position) %>%\n",
    "    mutate(total_smoothed=zoo::rollapply(total, smoothing_window, mean, align='center',fill=NA)) %>%\n",
    "    ungroup()\n",
    "\n",
    "  # Finally, restrict to x_steps points for plotting to speed things up\n",
    "  total_range = end - start\n",
    "  stepsize = ceiling(total_range / x_steps)\n",
    "  positions_to_show = seq(start, end, by=stepsize)\n",
    "  expanded_ranges = subset(expanded_ranges, position %in% positions_to_show)\n",
    "\n",
    "  # Make plots faceted by group\n",
    "  max_value = max(expanded_ranges$total_smoothed, na.rm = TRUE)\n",
    "\n",
    "  plot_object = ggplot() +\n",
    "    geom_bar(data=expanded_ranges, aes(position, total_smoothed), stat='identity', color='black', fill=\"#d3d3d3\", alpha=0.3) +\n",
    "    facet_wrap(~group, ncol=1, strip.position=\"right\") +\n",
    "    ylim(0, max_value * 1.4) +\n",
    "    xlab('Position (bp)') +\n",
    "    ylab(paste0('Normalized Signal (', smoothing_window, ' bp window)')) +\n",
    "    theme_classic()\n",
    "\n",
    "  # Add gene track using ensembl ID\n",
    "  gr <- GRanges(seqnames = str_replace(chrom, 'chr', ''), IRanges(start, end), strand = \"*\")\n",
    "  filters = AnnotationFilterList(GRangesFilter(gr), GenebiotypeFilter('protein_coding'))\n",
    "\n",
    "  genes = autoplot(ensdb, filters, names.expr = \"gene_name\")\n",
    "\n",
    "  genes_plot = genes@ggplot +\n",
    "    xlim(start, end) +\n",
    "    theme_classic()\n",
    "\n",
    "  # Now combine the two plots into one (using patchwork)\n",
    "  return(plot_object / (genes_plot + xlim(start, end)))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6ca68a27-42ab-43e8-b361-bbc715d97b06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 3"
      ],
      "text/latex": [
       "\\textbf{pdf:} 3"
      ],
      "text/markdown": [
       "**pdf:** 3"
      ],
      "text/plain": [
       "pdf \n",
       "  3 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "144e189d-0e62-41ab-9f68-6b3ca1bfeefe",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'position'. You can override using the\n",
      "`.groups` argument.\n",
      "Warning message:\n",
      "\"'GenebiotypeFilter' is deprecated.\n",
      "Use 'GeneBiotypeFilter' instead.\n",
      "See help(\"Deprecated\")\"\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Constructing graphics...\n",
      "\n",
      "Coordinate system already present. Adding new coordinate system, which will replace the existing one.\n",
      "\n",
      "Warning message:\n",
      "\"Removed 23 rows containing missing values (position_stack).\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "grouping_table = read.delim('subgroup_annotation_for_R.txt', sep='\\t')\n",
    "pdf(file=\"TFBS_HNF4A_ALB_III_20230125.pdf\")\n",
    "\n",
    "plot_region_pileups('chr4', 73415930, 73416346, 'fragments.tsv.gz', grouping=grouping_table, smoothing_window=10, x_steps = 10000)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50df951c-4703-48b8-9eca-2be41f2a1263",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
